lib/model/compose.dart

```diff
@@ -0,0 +1,179 @@
+const hashReplacements = {
+  "%": ".",
+  "(": ".28",
+  ")": ".29",
+  ".": ".2E",
+};
+
+// Corresponds to encodeHashComponent in Zulip web;
+// see web/shared/src/internal_url.js.
+String encodeHashComponent(String str) {
+  return Uri.encodeComponent(str)
+    .replaceAllMapped(RegExp(r'[%().]'), (Match m) => hashReplacements[m[0]!]!);
+}
+
+/// A URL to the given [Narrow], on `store`'s realm.
+///
+/// To include /near/{messageId} in the link, pass a non-null [nearMessageId].
+// Why take [nearMessageId] in a param, instead of looking for it in [narrow]?
+//
+// A reasonable question: the "near" part of a near link (e.g., for
+// quote-and-reply) takes the same form as other operator/operand pairs that we
+// represent with [ApiNarrowElement]s, like "/stream/48-mobile".
+//
+// But unlike those other elements, we choose not to give the "near" element
+// an [ApiNarrowElement] representation, because it doesn't have quite that role:
+// it says where to look in a list of messages, but it doesn't filter the list down.
+// In fact, from a brief look at server code, it seems to be *ignored*
+// if you include it in the `narrow` param in get-messages requests.
+// When you want to point the server to a location in a message list, you
+// you do so by passing the `anchor` param.
+Uri narrowLink(PerAccountStore store, Narrow narrow, {int? nearMessageId}) {
+  final apiNarrow = narrow.apiEncode();
+  final resultBuffer = StringBuffer('#narrow');
+  for (final element in apiNarrow) {
+    resultBuffer.write('/');
+    if (element.negated) {
+      resultBuffer.write('-');
+    }
+    // TODO(server-7) remove special-casing
+    if (element is ApiNarrowDm) {
+      final supportsOperatorDm = store.connection.zulipFeatureLevel! >= 177; // TODO(server-7)
+      final resolved = element.resolve(legacy: !supportsOperatorDm);
+      final operator = resolved.operator;
+      final operand = resolved.operand;
+      final operandSuffix = operand.length >= 3
+        ? 'group'
+        : (supportsOperatorDm ? 'dm' : 'pm'); // TODO(?) CZO actually gives a slugified full_name here…??
+      resultBuffer.write('$operator/${operand.join(',')}-$operandSuffix');
+      continue;
+    }
+
+    resultBuffer.write('${element.operator}/');
+
+    switch (element) {
+      case ApiNarrowStream():
+        final streamId = element.operand;
+        final name = store.streams[streamId]?.name ?? 'unknown';
+        final slugifiedName = encodeHashComponent(name.replaceAll(' ', '-'));
+        resultBuffer.write('$streamId-$slugifiedName');
+      case ApiNarrowTopic():
+        resultBuffer.write(encodeHashComponent(element.operand));
+      case ApiNarrowDm():
+        // do nothing; handled as special case above
+      case ApiNarrowMessageId():
+        resultBuffer.write(element.operand.toString());
+    }
+  }
+
+  if (nearMessageId != null) {
+    resultBuffer.write('/near/$nearMessageId');
+  }
+
+  return store.account.realmUrl.resolve(resultBuffer.toString());
```

Comment:
The relative URL in `resultBuffer` always goes purely in the fragment — it starts with `#` — so an equivalent version that's a bit simpler to know what it does is:
```suggestion
  return store.account.realmUrl.replace(fragment: resultBuffer.toString());
```

together with
```diff
-  final resultBuffer = StringBuffer('#narrow');
+  final resultBuffer = StringBuffer('narrow');
```

(And perhaps also rename the local to `fragmentBuffer`, or just `fragment`.)